derive/src/type_def.rs

```diff
@@ -14,19 +14,173 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use proc_macro2::TokenStream as TokenStream2;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
 use syn::{
-    self,
-    parse::Result,
+	self, parse::Result, parse_quote, punctuated::Punctuated, token::Comma, Data, DataEnum, DataStruct, DataUnion,
+	DeriveInput, Expr, ExprLit, Field, Fields, Ident, Lit, Variant,
 };
 
 pub fn generate(input: TokenStream2) -> TokenStream2 {
-    match generate_impl(input.into()) {
-        Ok(output) => output.into(),
-        Err(err) => err.to_compile_error().into(),
-    }
+	match generate_impl(input.into()) {
+		Ok(output) => output.into(),
+		Err(err) => err.to_compile_error().into(),
+	}
 }
 
-pub fn generate_impl(_input: TokenStream2) -> Result<TokenStream2> {
-	unimplemented!()
+pub fn generate_impl(input: TokenStream2) -> Result<TokenStream2> {
+	let mut ast: DeriveInput = syn::parse2(input)?;
+
+	// add bound
+	ast.generics.type_params_mut().for_each(|p| {
+		p.bounds.push(parse_quote!(_type_metadata::HasTypeId));
+	});
+
+	let ident = &ast.ident;
+	let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+
+	let def = match &ast.data {
+		Data::Struct(ref s) => generate_struct_def(s),
+		Data::Enum(ref e) => generate_enum_def(e),
+		Data::Union(ref u) => generate_union_def(u),
+	};
+
+	let has_type_def_impl = quote! {
+		impl #impl_generics _type_metadata::HasTypeDef for #ident #ty_generics #where_clause {
+			fn type_def() -> _type_metadata::TypeDef {
+				#def.into()
+			}
+		}
+	};
+
+	let mut renamed = String::from("_IMPL_HAS_TYPE_DEF_FOR_");
+	renamed.push_str(ident.to_string().trim_start_matches("r#"));
+	let dummy_const = Ident::new(&renamed, Span::call_site());
+	let output = quote! {
+		#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
+		const #dummy_const: () = {
+			#[allow(unknown_lints)]
+			#[cfg_attr(feature = "cargo-clippy", allow(useless_attribute))]
+			#[allow(rust_2018_idioms)]
+			use type_metadata as _type_metadata;
```

Comment:
Can you elaborate on why we use this alias?

Reply:
Same as here. If the context of metadata derive has already imported `type_metadata`, there would be a error complaining about re-importing. (But we cannot replied on the context to import `type_metadata` as it might be renamed in the context.)